File: 1986-minimum-number-of-work-sessions-to-finish-the-tasks-3.java
import java.util.*;

class LcMinimumNumberWorkSessionsSolution {
    /*
     * We can use backtracking way which would be faster than the other 2 solutions
     * We will track index of the task as well as a list of sessions that are
     * currently open
     * in the list of sessions, we will store the remainingSessionTime
     * So whenever we have a new task, we decide if we want to include in the
     * session based on whether we can fit the task. We will consider all
     * possibilities of fitting the task into each possible session as well as
     * creating a new session to house this new task
     */

    public static int result;

    public int minSessions(int[] tasks, int sessionTime) {
        Arrays.sort(tasks);
        for (int i = 0, j = tasks.length - 1; i < j; i++, j--) {
            int temp = tasks[i];
            tasks[i] = tasks[j];
            tasks[j] = temp;
        }
        int n = tasks.length;
        result = n;

        backtracking(tasks, sessionTime, 0, new ArrayList<Integer>());

        return result;
    }

    public void backtracking(int[] tasks, int sessionTime, int index, List<Integer> sessions) {
        if (sessions.size() >= result) {
            return;
        }

        if (index == tasks.length) {
            result = Math.min(result, sessions.size());
            return;
        }

        for (int i = 0; i < sessions.size(); i++) {
            int session = sessions.get(i);
            if (session >= tasks[index]) {

                sessions.set(i, session - tasks[index]);
                backtracking(tasks, sessionTime, index + 1, sessions);
                sessions.set(i, session);
            }
        }
        sessions.add(sessionTime - tasks[index]);
        backtracking(tasks, sessionTime, index + 1, sessions);
        sessions.remove(sessions.size() - 1);
    }

    public static void main(String[] args) {
        LcMinimumNumberWorkSessionsSolution sol = new LcMinimumNumberWorkSessionsSolution();

        System.out.println("output: " + sol.minSessions(new int[] { 1, 2, 3 }, 3) + ", expected: 2");
    }

}

File: 1986-minimum-number-of-work-sessions-to-finish-the-tasks-3.py
class Solution:
    def minSessions(self, tasks: list[int], sessionTime: int) -> int:
        '''
        We can use backtracking way which would be faster than the other 2 solutions
        We will track index of the task as well as a list of sessions that are currently open
        in the list of sessions, we will store the remainingSessionTime
        So whenever we have a new task, we decide if we want to include in the session based on whether we can fit the task. We will consider all possibilities of fitting the task into each possible session as well as creating a new session to house this new task
        '''

        tasks.sort(reverse=True)
        n = len(tasks)
        result = n
        
        def backtracking(i: int, sessions: list[int]):
            nonlocal result
            if len(sessions) >= result:
                return
            
            if i == n:
                result = min(result, len(sessions))
                return
            
            for j in range(len(sessions)):
                if sessions[j] >= tasks[i]:
                    sessions[j] -= tasks[i]
                    backtracking(i+1, sessions)
                    sessions[j] += tasks[i]
                    
            sessions.append(sessionTime - tasks[i])
            backtracking(i+1, sessions)
            sessions.pop()
            
        backtracking(0, [])
        return result


sol = Solution()
print(f'output: {sol.minSessions([1, 2, 3], 3)} expected: 2')
# print(f'output: {sol.minSessions([3, 1, 3, 1, 1], 8)} expected: 2')
# print(f'output: {sol.minSessions([1, 2, 3, 4, 5], 15)} expected: 1')
